pkg/zone/available_services.go

```diff
@@ -0,0 +1,149 @@
+package zone
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/prometheus/client_golang/prometheus"
+	"google.golang.org/protobuf/proto"
+
+	mesh_proto "github.com/kumahq/kuma/api/mesh/v1alpha1"
+	core_mesh "github.com/kumahq/kuma/pkg/core/resources/apis/mesh"
+	"github.com/kumahq/kuma/pkg/core/resources/manager"
+	"github.com/kumahq/kuma/pkg/core/user"
+	core_metrics "github.com/kumahq/kuma/pkg/metrics"
+	"github.com/kumahq/kuma/pkg/xds/cache/mesh"
+	xds_context "github.com/kumahq/kuma/pkg/xds/context"
+	"github.com/kumahq/kuma/pkg/xds/ingress"
+)
+
+type ZoneAvailableServicesTracker struct {
+	logger            logr.Logger
+	metric            prometheus.Summary
+	resManager        manager.ResourceManager
+	meshCache         *mesh.Cache
+	interval          time.Duration
+	ingressTagFilters []string
+	zone              string
+}
+
+func NewZoneAvailableServicesTracker(
+	logger logr.Logger,
+	metrics core_metrics.Metrics,
+	resManager manager.ResourceManager,
+	meshCache *mesh.Cache,
+	interval time.Duration,
+	ingressTagFilters []string,
+	zone string,
+) (*ZoneAvailableServicesTracker, error) {
+	metric := prometheus.NewSummary(prometheus.SummaryOpts{
+		Name:       "component_zone_available_services",
+		Help:       "Summary of available services tracker component interval",
+		Objectives: core_metrics.DefaultObjectives,
+	})
+	if err := metrics.Register(metric); err != nil {
+		return nil, err
+	}
+	return &ZoneAvailableServicesTracker{
+		logger:            logger,
+		metric:            metric,
+		resManager:        resManager,
+		meshCache:         meshCache,
+		interval:          interval,
+		ingressTagFilters: ingressTagFilters,
+		zone:              zone,
+	}, nil
+}
+
+func (t *ZoneAvailableServicesTracker) Start(stop <-chan struct{}) error {
+	t.logger.Info("starting")
+	ticker := time.NewTicker(t.interval)
+	ctx := user.Ctx(context.Background(), user.ControlPlane)
+
+	for {
+		select {
+		case <-ticker.C:
+			start := time.Now()
+			if err := t.updateZoneIngresses(ctx); err != nil {
+				t.logger.Error(err, "couldn't update ZoneIngress available services")
+			}
+			t.metric.Observe(float64(time.Since(start).Milliseconds()))
+		case <-stop:
+			t.logger.Info("stopping")
+			return nil
+		}
+	}
+}
+
+func (t *ZoneAvailableServicesTracker) NeedLeaderElection() bool {
+	return true
+}
+
+func availableServicesEqual(services []*mesh_proto.ZoneIngress_AvailableService, other []*mesh_proto.ZoneIngress_AvailableService) bool {
+	if len(services) != len(other) {
+		return false
+	}
+	for i := range services {
+		if !proto.Equal(services[i], other[i]) {
+			return false
+		}
+	}
+	return true
+}
+
+func (t *ZoneAvailableServicesTracker) getIngressExternalServices(
+	aggregatedMeshCtxs xds_context.AggregatedMeshContexts,
+) []*core_mesh.ExternalServiceResource {
+	var externalServices []*core_mesh.ExternalServiceResource
+
+	for _, mesh := range aggregatedMeshCtxs.Meshes {
+		if !mesh.ZoneEgressEnabled() {
```

Comment:
I think it's worth to drop a comment why we skipping. I know why but it might not be obvious for the reader